src/blr/cli/tagfastq.py

```diff
@@ -119,22 +101,44 @@ def parse_corrected_barcodes(open_file):
     return corrected_barcodes
 
 
-def parse_raw_barcodes(open_file):
-    """
-    Parse FASTA/FASTQ containing barcode sequences and return a dictionary with entry headers
-    pointing to a raw barcodes sequence.
-    :param open_file: dnaio odject.
-    :return: dict: entry headers pointing to a raw barcodes sequence
-    """
-    for barcode in tqdm(open_file, desc="Raw barcodes processed"):
-        header, _ = barcode.name.split(maxsplit=1)
-        yield {header: barcode.sequence}
+class BarcodeReader:
+    def __init__(self, filename):
+        self._cache = dict()
+        self._file = dnaio.open(filename, mode="r")
+        self.barcodes = self.parse()
+
+    def parse(self):
+        for barcode in tqdm(self._file, desc="Uncorrected barcodes processed"):
+            read_name, _ = barcode.name.split(maxsplit=1)
+            yield (read_name, barcode.sequence)
+
+    def get_barcode(self, read_name, maxiter=10):
+        if read_name in self._cache:
+            return self._cache.pop(read_name)
+
+        for barcode_read_name, barcode_sequence in islice(self.barcodes, maxiter):
+            # If read_name in next pair then parser lines are synced --> drop cache.
+            if read_name == barcode_read_name:
+                self._cache.clear()
+                return barcode_sequence
+
+            self._cache.update({barcode_read_name: barcode_sequence})
```

Comment:
This should simply be `self._cache[barcode_name] = barcode_sequence`

Reply:
Fixed!